### ISSUE TITLE
prefer-destructuring - Option to skip when destructurable variables  count less than N

### ISSUE THREAD

**Issue by bendtherules** [eslint/eslint#10743]
**What rule do you want to change?**
prefer-destructuring

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option (Not sure if it should have some non-zero default value)

**Please provide some example code that this change will affect:**

```js
const abc = this.state.abc;
// other non-assignment stuff
```

**What does the rule currently do for this code?**
Asks to convert to destructuring syntax, even if there is only one destructurable variable.
```js
const {abc} = this.state;
// other non-assignment stuff
```

**What will the rule do after it's changed?**
Keep it unchanged, for upto N allowed destructurable variables. (makes most sense when N = 1)

**Reasoning behind the suggestion:**

We use lot of code like this

```js
// Unfixed
const abc = this.state.abc;
// Fixed
const { abc } = this.state;
```

To me it doesn't make much sense to use destructuring just for one variables like this. First one is much more explicit without being repetitive (which is the case for multiple sequential assignments like shown below).

The rule definitely makes sense for multiple assignments like this - 

```js
const a = this.state.a;
const b = this.state.b;
const c = this.state.c;
```

because the fixed code is concise and clear -

```js
const { a, b, c } = this.state;
```

but for one variable, its hardly concise and looks too complex when it isn't.

Although it makes more sense to me to skip 1 variable situation, but we should probably keep it configurable as N.

**Comment by bendtherules** [comment]
Things to consider:

```js
const data = {};

// 1. Multiple variable declarators in same variable declaration
var a = data.a, b = data.b, c = 3;

// 2. Multiple sequential variable declarations (which can contain case 1 internally)
var d = data.d;
var e = data.e;


// 3. Non-sequential variable declarations
var f = data.f;
console.log(f);
var g = data.g;

// 4. Count across scopes - Shouldn't be added
var h = data.h
{
  var i = data.i;
}

```

**Comment by kaicataldo** [comment]
> To me it doesn't make much sense to use destructuring just for one variables like this. First one is much more explicit without being repetitive (which is the case for multiple sequential assignments like this).

> but for one variable, its hardly concise and looks too complex when it isn't

I'm not sure I agree with these statements. It really seems like a matter of personal taste/familiarity with destructuring syntax to me.

**Comment by bendtherules** [comment]
Sounds fair.
My rationale is that similar to 'as-needed' option (in arrow-parens, etc), destructuring should be optional when it is not needed (read: doesn't particularly serve the purpose of conciseness).

**Comment by kaicataldo** [comment]
Yeah, I just don't see much difference between `const abc = this.state.abc;` and `const { abc } = this.state;` as far as conciseness goes. Just my personal opinion though :) Can you elaborate on what specific cases this rule would warn and not warn about? I'm worried that it won't be able to warn on many cases because it just can't know for certain if it can or not, leading to an option that doesn't cover all the cases it should.

**Comment by bendtherules** [comment]
From what I can think of, there are two ways to count the number of possible destructurings (dest count) from a variable:

1. Sum up dest count for a group of sequential variable declaration statements to get local dest count for each destructuree variable. Whenever there is non-declaration statement, reset count of all destructuree variable. Also gets reset when moving to new scope.

Ex :
```js
// 1. Multiple variable declarators in same variable declaration
var a = data.a, b = 3, c = data.c; // dest count for `data` = 2
// ---------

// 2. Multiple sequential variable declarations (which can contain case 1 internally)
var d = data.d;
var e = data.e, f = data.f; // dest count for `data` = 3
// ---------

// 3. Non-sequential variable declarations
var f = data.f;  // dest count for `data` = 1
console.log(f);
var g = data.g;  // dest count for `data` = 1
// ---------

// 4. Count across scopes - Shouldn't be added
var h = data.h  // dest count for `data` = 1
{
  var i = data.i;  // dest count for `data` = 1
}
```

2. Sum up dest count for all variable declaration statements within a scope and reset across scopes
Ex :
```js
// 1. Multiple variable declarators in same variable declaration
var a = data.a, b = 3, c = data.c; // dest count for `data` = 2
// ---------

// 2. Multiple sequential variable declarations (which can contain case 1 internally)
var d = data.d;
var e = data.e, f = data.f; // dest count for `data` = 3
// ---------

// 3. Non-sequential variable declarations
var f = data.f; 
console.log(f);
var g = data.g;  // dest count for `data` = **2**
// ---------

// 4. Count across scopes - Shouldn't be added
var h = data.h  // dest count for `data` = 1
{
  var i = data.i;  // dest count for `data` = 1
}
```

I prefer the 2nd approach as it is more permissive (in terms of counting).

**Comment by ilyavolodin** [comment]
I think it's reasonable to require destructuring if multiple variables are being extracted from the same object. I think the message above should be augmented to include renaming as well.

**Comment by bendtherules** [comment]
Yes, I missed the other options in this rule. The examples given above are for `VariableDeclarator`.
Similarly, `enforceForRenamedProperties` should also modify the counting.

`AssignmentExpression` on the other hand is interesting as its not necessarily a statement. So, counting using rule 1 above makes no sense for this. Even more reason to count across the whole scope instead of sequential statements.

**Comment by bendtherules** [comment]
Added some thoughts on already destructured variables in https://github.com/eslint/eslint/issues/10743#issuecomment-415186659
